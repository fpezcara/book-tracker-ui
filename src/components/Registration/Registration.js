const Registration = () => {
  const handleSubmit = (e) => {
    e.preventDefault();
    const formData = new FormData(e.target);

    const formValues = {
      email: formData.get("email"),
      password: formData.get("password"),
      passwordConfirmation: formData.get("password-confirmation"),
    };

    console.log(formValues);
  };
  return (
    <section>
      <form onSubmit={handleSubmit}>
        <input type="email" name="email" placeholder="Email address" />
        <input type="password" name="password" placeholder="Password" />
        <input
          type="password"
          name="password-confirmation"
          placeholder="Password Confirmation"
        />
        <button type="submit">Register</button>
      </form>
    </section>
  );
};

export default Registration;
