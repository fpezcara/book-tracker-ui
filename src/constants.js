export const API_URL =
  process.env.NODE_ENV === "production"
    ? process.env.BOOK_TRACKER_API_PRODUCTION
    : process.env.REACT_APP_BOOK_TRACKER_API;
