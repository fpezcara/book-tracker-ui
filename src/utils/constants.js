export const API_URL =
  process.env.NODE_ENV === "production"
    ? process.env.REACT_APP_BOOK_TRACKER_API_PRODUCTION
    : process.env.REACT_APP_BOOK_TRACKER_API;

export const WS_URL =
  process.env.NODE_ENV === "production"
    ? process.env.REACT_APP_WSS_URL
    : process.env.REACT_APP_WS_URL;
