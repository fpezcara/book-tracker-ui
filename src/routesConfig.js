import Home from "./components/Home";
import Header from "./components/Header";
import AddBook from "./components/AddBook";
import NotFound from "./components/NotFound";
import Registration from "./components/Registration";
import Login from "./components/Login";

import { Outlet, Navigate } from "react-router-dom";

const routesConfig = [
  {
    path: "/",
    element: (
      <>
        <Header />
        <Outlet />
      </>
    ),
    errorElement: <NotFound />,
    children: [
      { path: "/", element: <Navigate to="/reading" /> },
      { path: "/404", element: <NotFound /> },
      { path: "/:name", element: <Home /> },
      { path: "/:name/add-book", element: <AddBook /> },
      { path: "/register", element: <Registration /> },
      { path: "/login", element: <Login /> },
    ],
  },
];

export default routesConfig;
