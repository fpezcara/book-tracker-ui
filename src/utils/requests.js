// import axios from "axios";
import { API_URL } from "./constants";

export const addBookToList = async (userId, listId, book) => {
  const res = await fetch(
    `${API_URL}/users/${userId}/lists/${listId}/add_book`,
    {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      credentials: "include",
      body: JSON.stringify({
        book: book,
      }),
    },
  );

  const data = await res.json();

  if (!res.ok) {
    throw data;
  }

  return data;
};

export const removeBookFromList = async (userId, listId, bookId) => {
  const res = await fetch(
    `${API_URL}/users/${userId}/lists/${listId}/remove_book`,
    {
      method: "DELETE",
      headers: {
        "Content-Type": "application/json",
      },
      credentials: "include",
      body: JSON.stringify({
        book_id: bookId,
      }),
    },
  );

  const data = await res.json();

  if (!res.ok) {
    throw res;
  }

  return data;
};

export const registerUser = async ({
  email_address,
  password,
  password_confirmation,
}) => {
  console.log("Sending request to:", `${API_URL}/users`);
  console.log("Request body:", {
    user: { email_address, password, password_confirmation },
  });
  const res = await fetch(`${API_URL}/users`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    credentials: "include",
    body: JSON.stringify({
      user: { email_address, password, password_confirmation },
    }),
  });

  const data = await res.json();

  if (!res.ok) {
    throw data;
  }

  return data;
};

export const logoutUser = async () => {
  const res = await fetch(`${API_URL}/session`, {
    method: "DELETE",
    headers: {
      "Content-Type": "application/json",
    },
    credentials: "include",
  });

  if (!res.ok) {
    throw res;
  }

  return res.json();
};

export const loginUser = async ({ email_address, password }) => {
  const res = await fetch(`${API_URL}/session`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    credentials: "include",
    body: JSON.stringify({ session: { email_address, password } }),
  });

  if (!res.ok) {
    throw res;
  }

  return res.json();
};
