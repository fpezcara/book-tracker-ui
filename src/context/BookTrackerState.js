import React, { useReducer, useEffect, useMemo, useState } from "react";
import BookTrackerContext from "./book-tracker-context";
import BookTrackerReducer from "./book-tracker-reducer";

import { API_URL } from "../utils/constants";
import Cookies from "js-cookie";

import {
  ADD_BOOK,
  DELETE_BOOK,
  CURRENT_BOOKLIST,
} from "./book-tracker-actions";

const BookTrackerState = ({ children }) => {
  const { Provider } = BookTrackerContext;
  const userId = Cookies.get("userId");
  // todo: rename to bookLists
  const [lists, setLists] = useState([]);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState({});

  const updateCurrentBookList = (selectedBookList) => {
    dispatch({
      type: CURRENT_BOOKLIST,
      payload: selectedBookList,
    });
  };

  const addBook = (book) => {
    dispatch({
      type: ADD_BOOK,
      payload: book,
    });
  };
  const deleteBook = (book) => {
    dispatch({
      type: DELETE_BOOK,
      payload: book,
    });
  };
  const initialState = useMemo(() => {
    return {
      bookLists: [
        { id: 1, listUrl: "reading", listName: "Reading", books: [] },
        { id: 2, listUrl: "finished", listName: "Finished", books: [] },
        { id: 3, listUrl: "wishlist", listName: "Wishlist", books: [] },
      ],
      currentBookList: "reading",
    };
  }, []);

  const [state, dispatch] = useReducer(BookTrackerReducer, initialState);

  useEffect(() => {
    if (JSON.parse(localStorage.getItem("state"))) {
      dispatch({
        type: "init_stored",
        payload: JSON.parse(localStorage.getItem("state")),
      });
    }
  }, []);

  useEffect(() => {
    try {
      userId &&
        fetch(`${API_URL}/users/${userId}/lists`, {
          method: "GET",
          headers: {
            "Content-Type": "application/json",
          },
          credentials: "include",
        })
          .then((response) => response.json())
          .then((data) => {
            console.log("Fetched data:", data);
            setLists(data);
            setLoading(false);
          });
    } catch (error) {
      console.error("Error fetching book lists:", error);
      setError(error);
      setLoading(false);
    }

    if (state !== initialState) {
      localStorage.setItem("state", JSON.stringify(state));
    }
  }, [state, initialState, userId]);

  const values = {
    state,
    addBook,
    deleteBook,
    updateCurrentBookList,
    lists,
    loading,
    error,
  };

  return <Provider value={values}>{children}</Provider>;
};

export default BookTrackerState;
