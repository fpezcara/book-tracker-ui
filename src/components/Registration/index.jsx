import { useState } from "react";
import { useNavigate } from "react-router";
import { registerUser } from "../../utils/requests";

import Cookies from "js-cookie";

import {
  AuthenticationContainer,
  EmailTakenErrorMessage,
  RegistrationForm,
} from "../../styles/Authentication.style";

const Registration = () => {
  const navigate = useNavigate();
  const [errorMessage, setErrorMessage] = useState("");

  const handleSubmit = async (e) => {
    e.preventDefault();
    const formData = new FormData(e.target);

    const formValues = {
      email_address: formData.get("email"),
      password: formData.get("password"),
      password_confirmation: formData.get("password-confirmation"),
    };

    registerUser({ ...formValues })
      .then((response) => {
        Cookies.set("userId", response?.user_id);
        Cookies.get("currentBookList") ||
          Cookies.set("currentBookList", "reading");
        navigate("/");
      })
      .catch((error) => {
        if (
          error.response?.status === 400 &&
          error.response?.data?.message.includes(
            "Email address has already been taken",
          )
        ) {
          setErrorMessage(
            "Email address has already been taken. Please try again.",
          );
        } else {
          console.error(error);
        }
      });
  };

  return (
    <AuthenticationContainer>
      <div className="wrapper">
        <h2>Register</h2>
        <RegistrationForm onSubmit={handleSubmit}>
          <input type="email" name="email" placeholder="Email address" />
          <input type="password" name="password" placeholder="Password" />
          <input
            type="password"
            name="password-confirmation"
            placeholder="Password Confirmation"
          />
          {errorMessage && (
            <EmailTakenErrorMessage>{errorMessage}</EmailTakenErrorMessage>
          )}
          <button type="submit">Register</button>
        </RegistrationForm>
      </div>
    </AuthenticationContainer>
  );
};

export default Registration;
