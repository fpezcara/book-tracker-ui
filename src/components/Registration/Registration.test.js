import React from "react";
import { render, screen, waitFor, fireEvent } from "@testing-library/react";
import { RouterProvider, createMemoryRouter } from "react-router";
import routesConfig from "../../routesConfig";
import Cookies from "js-cookie";
import BookTrackerState from "../../context/BookTrackerState";
import { API_URL } from "../../utils/constants";
import Registration from "./";

jest.mock("../../utils/constants", () => ({
  API_URL: "book-tracker-api.com",
}));

const submitForm = (bodyPayload) => {
  // Fill out the form
  fireEvent.change(screen.getByPlaceholderText("Email address"), {
    target: { value: bodyPayload.email_address },
  });
  fireEvent.change(screen.getByPlaceholderText("Password"), {
    target: { value: bodyPayload.password },
  });
  fireEvent.change(screen.getByPlaceholderText("Password Confirmation"), {
    target: { value: bodyPayload.password_confirmation },
  });

  // Submit the form
  fireEvent.click(screen.getByRole("button", { name: /register/i }));
};

describe("Registration", () => {
  const bodyPayload = {
    email_address: "test@example.com",
    password: "password123",
    password_confirmation: "password123",
  };

  const userId = 9;

  beforeEach(() => {
    fetch.resetMocks();
  });

  afterEach(() => {
    Cookies.remove("userId");
  });

  it("renders correctly", async () => {
    const router = createMemoryRouter(routesConfig, {
      initialEntries: ["/register"],
    });

    const { container } = render(
      <BookTrackerState>
        <RouterProvider router={router}>
          <Registration />
        </RouterProvider>
      </BookTrackerState>,
    );

    expect(container).toMatchSnapshot();
  });

  it("submits the form and sets cookie on success", async () => {
    const bodyPayload = {
      email_address: "test@example.com",
      password: "password123",
      password_confirmation: "password123",
    };
    fetch.mockResponseOnce(JSON.stringify({ user_id: userId }));

    const router = createMemoryRouter(routesConfig, {
      initialEntries: ["/register"],
    });

    render(
      <BookTrackerState>
        <RouterProvider router={router}>
          <Registration />
        </RouterProvider>
      </BookTrackerState>,
    );

    submitForm(bodyPayload);

    await waitFor(() => {
      expect(router.state.location.pathname).toBe("/reading");
    });

    await waitFor(() => {
      expect(Cookies.get("userId")).toBe(userId.toString());
    });

    await waitFor(() => {
      expect(Cookies.get("currentBookList")).toBe("reading");
    });

    expect(fetch).toHaveBeenCalledWith(
      `${API_URL}/users`,
      expect.objectContaining({
        method: "POST",
        headers: {
          "Content-Type": "application/json",
        },
        credentials: "include",
        body: JSON.stringify({
          user: bodyPayload,
        }),
      }),
    );
  });

  it("displays error message in form if email address has already been taken", async () => {
    const router = createMemoryRouter(routesConfig, {
      initialEntries: ["/register"],
    });

    fetch.mockRejectOnce(
      new Error("Validation failed: Email address has already been taken"),
    );

    render(
      <BookTrackerState>
        <RouterProvider router={router}>
          <Registration />
        </RouterProvider>
      </BookTrackerState>,
    );

    submitForm(bodyPayload);

    await waitFor(() => {
      expect(
        screen.getByText(/Please choose a different email./i),
      ).toBeInTheDocument();
    });
  });

  it("displays error message in form if password & password_confirmation do not match", async () => {
    fetch.mockRejectOnce(
      new Error(
        "Validation failed: Password confirmation doesn't match Password",
      ),
    );

    const router = createMemoryRouter(routesConfig, {
      initialEntries: ["/register"],
    });

    render(
      <BookTrackerState>
        <RouterProvider router={router}>
          <Registration />
        </RouterProvider>
      </BookTrackerState>,
    );

    submitForm(bodyPayload);

    await waitFor(() => {
      expect(screen.getByText(/do not match./i)).toBeInTheDocument();
    });
  });

  it("displays generic error message on error ", async () => {
    fetch.mockRejectOnce(new Error());

    const router = createMemoryRouter(routesConfig, {
      initialEntries: ["/register"],
    });

    render(
      <BookTrackerState>
        <RouterProvider router={router}>
          <Registration />
        </RouterProvider>
      </BookTrackerState>,
    );

    submitForm(bodyPayload);

    await waitFor(() => {
      expect(
        screen.getByText(/an error occurred during registration./i),
      ).toBeInTheDocument();
    });
  });
});
