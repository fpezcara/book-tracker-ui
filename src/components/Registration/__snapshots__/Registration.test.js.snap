// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Registration renders correctly 1`] = `
<div>
  <header
    class="sc-jJAtPt FxFST"
  >
    <div
      class="authentication"
    >
      <a
        class="link home"
        data-discover="true"
        href="/undefined"
      >
        Home
      </a>
      <div
        class="authLinks"
      >
        <a
          class="link"
          data-discover="true"
          href="/register"
        >
          Register
        </a>
        <a
          class="link"
          data-discover="true"
          href="/login"
        >
          Login
        </a>
      </div>
    </div>
    <div>
      <h1>
        Book Tracker
      </h1>
    </div>
  </header>
  <section
    class="sc-etfXYe bTHtvQ"
  >
    <div
      class="wrapper"
    >
      <h2>
        Register
      </h2>
      <form
        class="sc-dFaThA kfDRsn"
      >
        <input
          name="email"
          placeholder="Email address"
          type="email"
        />
        <input
          name="password"
          placeholder="Password"
          type="password"
        />
        <input
          name="password-confirmation"
          placeholder="Password Confirmation"
          type="password"
        />
        <button
          type="submit"
        >
          Register
        </button>
      </form>
    </div>
  </section>
</div>
`;
